ListObjects Metadata Caching

klauspost opened the pull request:
## Motivation and Context

Design: https://gist.github.com/klauspost/025c09b48ed4a1293c917cecfabdf21c

## Improvements

* Cross-server caching and listing will use the same data across servers and requests.
* Lists can be arbitrarily resumed at constant speed.
* Metadata for all files scanned are stored for streaming retrieval.
* The existing bloom filters controlled by the crawler is used for validating caches.
* Concurrent requests for the same data (or parts of it) will not spawn additional walkers.
* Listing a subdirectory of an existing recursive cache will use the cache.
* All listing operations are fully streamable so the number of objects in a bucket no longer dictates the amount of memory.
* Listings can be handled by any server within the cluster.
* Caches are cleaned up when out of date or superseded by a more recent one.

## How to test this PR?

List Objects. Concurrently for example.

Cached entries are considered valid if:

A) If the request contains the list id (as a marker or continuation token), or
B) A listing is currently running on the data being requested, or
C) The bloom filter is currently clean for the requested path.

If the bloom filter is dirty and the listing has finished a new listing will start.

Lists are maintained at least until it is more than 1 hour since the ID of the list has been handed out. Usually much longer unless a newer list of the same data is found.

Recursive listings can be used for non-recursive listings or listings of subpaths. Only the path of the sublisting needs to be clean for the cache to be used.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] API Upgrade required

## Checklist:
- [x] Unit tests for `metacache` methods.
- [x] Implement ListObjects and ListObjectsV2 to use the same data. Currently only ListObjectVersions is using it.
- [x] Remove debug printfs.
- [x] Check error conditions/failures better.
- [ ] Testing ofc.

kerneltime commented on cmd/xl-storage.go:
```diff
@@ -1960,7 +1960,7 @@ func deleteFile(basePath, deletePath string, recursive bool) error {
 
 // DeleteFile - delete a file at path.
-func (s *xlStorage) DeleteFile(ctx context.Context, volume string, path string) (err error) {
+func (s *xlStorage) DeleteFile(ctx context.Context, volume string, path string, recursive bool) (err error) {
```
Nit: Rename to delete path?

klauspost replied:
Good. Renaming it to `Delete`

vadmeste commented on cmd/data-update-tracker.go:
```diff
@@ -168,6 +168,43 @@ func (d *dataUpdateTracker) current() uint64 {
+	}
+
+	idx := d.Current.idx - 1
+	for {
```
Maybe 'for idx >= 0 {' ?

klauspost replied:
It is a uint64 :D

I get what you mean. `d.History.find` will return nil, but we might as well handle it.

vadmeste commented on cmd/erasure-healing.go:
```diff
@@ -747,14 +769,18 @@ func (er erasureObjects) HealObject(ctx context.Context, bucket, object, version
 		if m.Erasure.DataBlocks == 0 {
 			writeQuorum = getWriteQuorum(len(storageDisks))
 		}
-		if !opts.DryRun && opts.Remove {
```
opts.Remove is still needed because we are using --remove in mc admin heal

vadmeste commented on cmd/storage-rest-common.go:
```diff
@@ -17,7 +17,7 @@
 
 const (
-	storageRESTVersion       = "v20" // Re-implementation of storage layer
+	storageRESTVersion       = "v21" // Add dir listing.
```
and add a recursive flag to Delete()

klauspost replied:
Yeah, not sure much value this has, but sure.

harshavardhana commented on cmd/erasure-common.go:
```diff
@@ -47,10 +49,17 @@ func (er erasureObjects) getOnlineDisks() (newDisks []StorageAPI) {
 			if disks[i-1] == nil {
+				if printDebug {
+					fmt.Println("getOnlineDisks: disk ", i-1, "was nil")
+				}
```
These can be removed

harshavardhana commented on cmd/erasure-object.go:
```diff
@@ -703,10 +707,9 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 }
 
 func (er erasureObjects) deleteObjectVersion(ctx context.Context, bucket, object string, writeQuorum int, fi FileInfo) error {
+	defer ObjectPathUpdated(path.Join(bucket, object))
```
```suggestion
	defer ObjectPathUpdated(pathJoin(bucket, object))
```

To preserve the object `/` at at the end

klauspost replied:
It doesn't really matter. It is always stripped anyway. Bloom filter doesn't distinguish.

vadmeste commented on cmd/notification.go:
```diff
@@ -507,6 +509,124 @@ func (sys *NotificationSys) updateBloomFilter(ctx context.Context, current uint6
+		client := client
+		g.Go(func() error {
+			serverBF, err := client.cycleServerBloomFilter(ctx, req)
+			if false && intDataUpdateTracker.debug {
```
s/false//

klauspost replied:
Too noisy.

vadmeste commented on cmd/notification.go:
```diff
@@ -507,6 +509,124 @@ func (sys *NotificationSys) updateBloomFilter(ctx context.Context, current uint6
+	return bf, nil
+}
+
+// updateBloomFilter will cycle all servers to the current index and
```
Can you update the comment here?

vadmeste reviewed (commented):
Some comments, did not finish reviewing yet

kerneltime commented on cmd/test-utils_test.go:
```diff
@@ -90,15 +89,15 @@ func init() {
 
 	// Disable printing console messages during tests.
-	color.Output = ioutil.Discard
+	// color.Output = ioutil.Discard
```
Is this due to WIP or do you plan to check this in?

klauspost replied:
TBH I find it extremely frustrating that debug logging is killed. You know what let's tie it up to `testing.Verbose`.

kerneltime commented on cmd/storage-rest-server.go:
```diff
@@ -787,6 +794,161 @@ func waitForHTTPResponse(respBody io.Reader) (io.Reader, error) {
+// An optional error can be sent which will be picked as text only error,
+// without its original type by the receiver.
+// waitForHTTPStream should be used to the receiving side.
+func streamHTTPResponse(w http.ResponseWriter) *httpStreamResponse {
```
Why does `keepalive` not serve the purpose of keeping the connection active?

klauspost replied:
Great Q. Let's look at it from the callers position.

This is how it is used:
```
	writer := streamHTTPResponse(w)
	writer.CloseWithError(s.storage.WalkDir(r.Context(), WalkDirOptions{...}, writer))
```
Using the `keepHTTPResponseAlive` it would look something like this:
```
	done := keepHTTPResponseAlive(w)
	done(nil)
	err := s.storage.WalkDir(r.Context(), WalkDirOptions{Bucket: volume, BaseDir: dirPath, Recursive: recursive}, w)
````

So, first of all `keepHTTPResponseAlive` is not really doing anything. Since we are streaming to the response writer `w` we need to close it right away so we don't race, making it useless.

We can do something else:
```
	done := keepHTTPResponseAlive(w)
	var tmp bytes.Buffer
	err := s.storage.WalkDir(r.Context(), WalkDirOptions{...}, &tmp)
	if err == io.EOF {
		err = nil
	}
	done(err)
	if err == nil {
		w.Write(tmp.Bytes())
	}
```
First of all this is not streaming anymore since we collect everything in a buffer. If the call (as here) can take several minutes and collect huge result this is bad in both ways.

Another thing is that we can not distinguish between `io.EOF` and nil.

Compare this to the cleanliness of this:

```
	writer := streamHTTPResponse(w)
	writer.CloseWithError(s.storage.WalkDir(r.Context(), WalkDirOptions{...}, writer))
```

Fully streaming, with keepalive padding and allows to return an error at any point in the stream.

kerneltime commented on cmd/peer-rest-server.go:
```diff
@@ -656,6 +657,51 @@ func (s *peerRESTServer) CycleServerBloomFilterHandler(w http.ResponseWriter, r
+	}
+	ctx := newContext(r, w, "GetMetacacheListing")
+
+	var req listPathOptions
```
Nit: `req` => `opts`?

kerneltime commented on cmd/metacache-bucket.go:
```diff
@@ -0,0 +1,427 @@
+	// Internal state
+	mu        sync.RWMutex `msg:"-"`
+	updated   bool         `msg:"-"`
+	transient bool         `msg:"-"`
```
Please document what transient implies and used for.

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -424,6 +424,14 @@ func extractAPIVersion(r *http.Request) string {
 
+func methodNotAllowedHandler(w http.ResponseWriter, r *http.Request) {
+	writeErrorResponseString(r.Context(), w, APIError{
+		Code:           "XMinioPeerVersionMismatch",
```
This code is used for all handlers and errors, you should perhaps add context here per API? such as `XMinioAdminVersionMismatch`,..

`XMinioS3APIMismatch` ?

harshavardhana commented on cmd/metacache-marker.go:
```diff
@@ -0,0 +1,70 @@
+	tags := strings.Split(tag, ",")
+	for _, tag := range tags {
+		kv := strings.Split(tag, ":")
+		if len(kv) < 2 {
```
shouldn't we fail for invalid markers?

klauspost replied:
No need to really. The worst that will happen is that the cache will not get automatically reused.

harshavardhana commented on cmd/metacache-marker.go:
```diff
@@ -0,0 +1,70 @@
+const markerTagVersion = "v1"
+
+// parseMarker will parse a marker possibly encoded with encodeMarker
+func parseMarker(s string) (marker, uuid string) {
```
`if s == "" { return s, "" }` - is enough? 

what if the marker doesn't conform or is mutated by the client what do we respond? - should we enforce this such that client indeed responds properly without manipulating the marker?

klauspost replied:
The marker can be used for specifying an object offset, so in those cases we just return the unmodified marker (code right below).

If we cannot parse it, we simply use whatever they gave as an offset.

harshavardhana commented on cmd/metacache-marker.go:
```diff
@@ -0,0 +1,70 @@
+	if strings.ContainsAny(uuid, "[:,") {
+		logger.LogIf(context.Background(), fmt.Errorf("encodeMarker: uuid %s contained invalid characters", uuid))
+	}
+	return fmt.Sprintf("%s[minio_cache:%s,id:%s]", marker, markerTagVersion, uuid)
```
It would be nice if we can `base64` encode this? because for clients it would avoid the extra effort to URL encode them when sent as `marker=%s` would make it easier to consume like continuation token?

harshavardhana replied:
Added benefit is same format shall be used for continuation token as well.

klauspost replied:
Object names (markers) are already URL encoded since it has slashes for instance. Not really sure what the benefit would be, since you'd need to encode it anyway.

This is used as continuation token already.

Individual values can be encoded however they like, if we should need to add that.

harshavardhana reviewed (commented):
Some review - will test over the weekend.

klauspost committed:
ListObjects Metadata Caching

Design: https://gist.github.com/klauspost/025c09b48ed4a1293c917cecfabdf21c

Improvements:

* Cross-server caching and listing will use the same data across servers and requests.
* Lists can be arbitrarily resumed at constant speed.
* Metadata for all files scanned are stored for streaming retrieval.
* The existing bloom filters controlled by the crawler is used for validating caches.
* Concurrent requests for the same data (or parts of it) will not spawn additional walkers.
* Listing a subdirectory of an existing recursive cache will use the cache.
* All listing operations are fully streamable so the number of objects in a bucket no longer dictates the amount of memory.
* Listings can be handled by any server within the cluster.
* Caches are cleaned up when out of date or superseded by a more recent one.
